graphql/index.ts

```diff
@@ -1,65 +1,53 @@
 import { composeWithMongoose } from 'graphql-compose-mongoose';
 import { GQC } from 'graphql-compose';
 import { UserModel } from '../models/UserProfile';
+import { some as promiseSome } from 'bluebird';
 
 const UserTC = composeWithMongoose(UserModel, {});
 
-function requireToken(resolvers) {
-  return Object.keys(resolvers).reduce((acc, key) => {
-    return {
-      ...acc,
-      [key]: resolvers[key].wrapResolve(next => resolverParameters => {
-        if (!resolverParameters.context.jwt.valid) {
-          throw new Error('You must provide valid token');
-        }
-        return next(resolverParameters);
-      }),
-    };
-  }, {});
+async function validToken({ context }) {
+  if (!context.jwt.valid) {
+    throw new Error('You must provide valid token');
+  } else {
+    return true;
+  }
 }
 
-function requireSameUserMongoId(resolvers) {
-  return Object.keys(resolvers).reduce((acc, key) => {
-    return {
-      ...acc,
-      [key]: resolvers[key].wrapResolve(next => async resolverParameters => {
-        const _id =
-          resolverParameters.args._id || resolverParameters.args.record._id;
-        const egoId = await UserModel.findOne({ _id }).then(
-          user => user.ego_id,
-        );
+async function isSelf({ args, context }) {
+  const _id = args._id || args.record._id;
+  const egoId = await UserModel.findOne({ _id }).then(user => user.ego_id);
+
+  if (args.record && args.record.ego_id !== egoId) {
+    throw new Error("You can't change your ego id");
+  } else if (`${egoId}` !== `${context.jwt.sub}`) {
+    throw new Error("You can't edit someone elses profile");
+  } else {
+    return true;
+  }
+}
 
-        if (
-          resolverParameters.args.record &&
-          resolverParameters.args.record.ego_id !== egoId
-        ) {
-          throw new Error("You can't change your ego id");
-        } else if (
-          `${egoId}` !== `${resolverParameters.context.jwt.payload.sub}`
-        ) {
-          throw new Error("You can't edit someone elses profile");
-        } else {
-          return next(resolverParameters);
-        }
+function restrict(resolver, ...restrictions) {
+  return resolver.wrapResolve(next => resolverParameters => {
+    return Promise.all(
+      restrictions.map(restriction => {
+        return promiseSome(
```

Comment:
Does node's native promise not have Promise.race? Or would that no be sufficient here?

Reply:
Race takes rejects and resolves. so if the first one is a reject than it will fail.

Reply:
Ah, makes sense